sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,372 @@
+        flush();
+      }
+    } catch (Throwable t) {
+      throw new RuntimeException(t);
+    }
+  }
+
+  /** Check if the flush thread failed with an exception. */
+  private void checkFlushThreadException() throws IOException {
+    if (timeLimit > 0 && flushFuture.isDone()) {
+      try {
+        flushFuture.get();
+        throw new IOException("Periodic flushing thread finished unexpectedly.");
+      } catch (ExecutionException ee) {
+        unwrapExecutionException(ee);
+      } catch (CancellationException ce) {
+        throw new IOException(ce);
+      } catch (InterruptedException ie) {
+        Thread.currentThread().interrupt();
+        throw new IOException(ie);
+      }
+    }
+  }
+
+  private void unwrapExecutionException(ExecutionException ee) throws IOException {
+    // the cause is always RuntimeException
+    RuntimeException re = (RuntimeException) ee.getCause();
+    if (re.getCause() instanceof IOException) {
+      throw (IOException) re.getCause();
+    } else {
+      throw new IOException(re.getCause());
+    }
+  }
+
+  private static int getSizeLimit(PipelineOptions options) {
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(DATA_BUFFER_SIZE_LIMIT)) {
+        return Integer.parseInt(experiment.substring(DATA_BUFFER_SIZE_LIMIT.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT_BYTES;
+  }
+
+  private static long getTimeLimit(PipelineOptions options) {
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(DATA_BUFFER_TIME_LIMIT_MS)) {
+        return Long.parseLong(experiment.substring(DATA_BUFFER_TIME_LIMIT_MS.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT_TIME_MS;
+  }
+
+  private class Receiver<T> implements FnDataReceiver<T> {
+    private final ByteString.Output output;
+    private final Coder<T> coder;
+    private long byteCount;
+    private long elementCount;
+
+    public Receiver(Coder<T> coder) {
+      this.output = ByteString.newOutput();
+      this.coder = coder;
+      this.byteCount = 0L;
+    }
+
+    @Override
+    public void accept(T input) throws Exception {
+      int size = output.size();
+      coder.encode(input, output);
+      if (output.size() - size == 0) {
+        output.write(0);
+      }
+      final long delta = (long) output.size() - size;
+      totalByteCounter.add(delta);
+      byteCount += delta;
+      elementCount += 1;
+      if (totalByteCounter.longValue() > sizeLimit) {
+        flush();
+      }
+    }
+
+    public ByteString.Output getOutput() {
+      return output;
+    }
+
+    public void reset() {
+      this.byteCount = 0L;
+      this.elementCount = 0L;
+      this.output.reset();
+    }
```

Comment:
You might want to differentiate `reset` for `flush` and for `finish the bundle` use cases since in the former you just want to reset the output bytestring to keep the element counts and for the latter you want to reset all.